File: другое/тинькофф.аисд/3/D.ts
class Heap {
  array: number[];
  constructor() {
    this.array = [];
  }

  add(item: number) {
    this.array.push(item);
    this.up(this.array.length - 1);
  }

  up(index: number) {
    let parentIndex = Math.floor((index - 1) / 2);
    while (index > 0 && this.array[parentIndex] < this.array[index]) {
      [this.array[parentIndex], this.array[index]] = [
        this.array[index],
        this.array[parentIndex],
      ];
      index = parentIndex;
      parentIndex = Math.floor((index - 1) / 2);
    }
  }

  remove() {
    const max = this.array[0];
    this.array[0] = this.array.pop()!;
    this.down(0);
    return max;
  }

  down(index: number) {
    let cur = index;
    let left = 2 * index + 1;
    let right = 2 * index + 2;
    let ind = index;
    if (
      left < this.array.length &&
      this.array[left] > this.array[cur]
    ) {
      cur = left;
    }

    if (
      right < this.array.length &&
      this.array[right] > this.array[cur]
    ) {
      cur = right;
    }

    if (cur !== index) {
      [this.array[index], this.array[cur]] = [
        this.array[cur],
        this.array[index],
      ];
      this.down(cur);
    }
  }
}

File: другое/тинькофф.аисд/3/D.py
class Heap:
    def __init__(self):
        self.array = []

    def add(self, item):
        self.array.append(item)
        self.up(len(self.array) - 1)

    def up(self, index):
        parent_index = (index - 1) // 2
        while index > 0 and self.array[parent_index] < self.array[index]:
            self.array[parent_index], self.array[index] = self.array[index], self.array[parent_index]
            index = parent_index
            parent_index = (index - 1) // 2

    def remove(self):
        max_value = self.array[0]
        self.array[0] = self.array[-1]
        self.array.pop()
        self.down(0)
        return max_value

    def down(self, index):
        cur = index
        left = 2 * index + 1
        right = 2 * index + 2
        if left < len(self.array) and self.array[left] > self.array[cur]:
            cur = left
        if right < len(self.array) and self.array[right] > self.array[cur]:
            cur = right
        if cur != index:
            self.array[index], self.array[cur] = self.array[cur], self.array[index]
            self.down(cur)


n = int(input())
ans = []
heap = Heap()
for _ in range(n):
    line = input().split()
    if line[0] == "0":
        heap.add(int(line[1]))
    else:
        print(heap.remove())
#
# for i in range(len(ans)):
#     sys.stdout.write(str(ans[i]) + '\n')
